pkg/cfg/config_test.go

```diff
@@ -1,163 +1,259 @@
+func (s *ConfigTestSuite) SetupTest() {
+	s.environment = map[string]string{}
 
-	assert.Len(t, keys, 2)
+	options := []cfg.Option{
+		cfg.WithErrorHandlers(s.errorHandler),
+		cfg.WithEnvKeyReplacer(strings.NewReplacer(".", "_")),
+		cfg.WithSanitizers(cfg.TimeSanitizer),
+	}
+
+	s.config = cfg.NewWithInterfaces(s.environmentMock)
+
+	if err := s.config.Option(options...); err != nil {
+		s.FailNow("con not apply config options", err.Error())
+	}
 }
 
-func TestConfig_IsSet(t *testing.T) {
-	config := getNewTestableConfig(baseSettings, map[string]string{})
+func (s *ConfigTestSuite) applyOptions(options ...cfg.Option) {
+	if err := s.config.Option(options...); err != nil {
+		s.FailNow("con not apply config options", err.Error())
+	}
+}
 
-	assert.True(t, config.IsSet("a"))
-	assert.True(t, config.IsSet("n.c"))
-	assert.False(t, config.IsSet("b"))
+func (s *ConfigTestSuite) errorHandler(err error, msg string, args ...interface{}) {
+	s.FailNow(err.Error(), fmt.Sprintf(msg, args...))
 }
 
-func TestConfig_Get(t *testing.T) {
-	config := getNewTestableConfig(baseSettings, map[string]string{})
+func (s *ConfigTestSuite) environmentMock(key string) (string, bool) {
+	if value, ok := s.environment[key]; ok {
+		return value, true
+	}
 
-	expectedMap := map[string]interface{}{
-		"c": 3,
+	return "", false
+}
+
+func (s *ConfigTestSuite) setupConfigValues(values map[string]interface{}) {
+	if err := s.config.Option(cfg.WithConfigMap(values)); err != nil {
+		s.FailNow("can not setup config values", err.Error())
 	}
+}
 
-	assert.Equal(t, 1, config.Get("a"))
-	assert.Equal(t, expectedMap, config.Get("n"))
+func (s *ConfigTestSuite) setupEnvironment(values map[string]string) {
+	s.environment = values
 }
 
-func TestConfig_GetBool(t *testing.T) {
-	config := getNewTestableConfig(map[string]interface{}{
-		"b": true,
-	}, map[string]string{})
+func (s *ConfigTestSuite) TestConfig_AllKeys() {
+	s.setupConfigValues(map[string]interface{}{
+		"i": 1,
+		"s": "string",
+	})
 
-	assert.True(t, config.GetBool("b"))
+	keys := s.config.AllKeys()
+	s.Len(keys, 2)
 }
 
-func TestConfig_GetDuration(t *testing.T) {
-	config := getNewTestableConfig(map[string]interface{}{
-		"d": time.Second,
-	}, map[string]string{})
+func (s *ConfigTestSuite) TestConfig_IsSet() {
+	s.setupConfigValues(map[string]interface{}{
+		"i": 1,
+		"ms": map[string]interface{}{
+			"b": true,
+		},
+	})
 
-	assert.Equal(t, time.Second, config.GetDuration("d"))
+	s.True(s.config.IsSet("i"))
+	s.True(s.config.IsSet("ms.b"))
+	s.False(s.config.IsSet("missing"))
 }
 
-func TestConfig_GetInt(t *testing.T) {
-	config := getNewTestableConfig(map[string]interface{}{
+func (s *ConfigTestSuite) TestConfig_Get() {
+	s.setupConfigValues(map[string]interface{}{
 		"i": 1,
-	}, map[string]string{})
+		"ms": map[string]interface{}{
+			"b": true,
+		},
+	})
+
+	expectedMap := map[string]interface{}{
+		"b": true,
+	}
+
+	s.Equal(1, s.config.Get("i"))
+	s.Equal(expectedMap, s.config.Get("ms"))
```

Comment:
It would also be interesting to add a test case that we get the value from the config even if we supply a different default:

```golang
	s.Equal(1, s.config.Get("i"), "2")
	s.Equal(expectedMap, s.config.Get("ms"), map[string]interface{}{
		"c": false,
	})
```

Reply:
added the test